File: net/et_runner.py
"""et_runner.py — CLIENT side: run one .pte on the ExecuTorch runtime.

Thin and torch-light (it does import torch for tensor I/O, but does NO generation,
no z3, no export). This is the exact work a mobile phone client mirrors natively:
load the program, execute the forward method on the provided inputs, return outputs.

`run_pte(pte_bytes, inputs)` returns a list of output tensors. A hard kernel failure
in the ExecuTorch runtime is a NATIVE abort that takes the process down — which is
why the client runs each job in a respawnable subprocess and the server treats a
dropped connection as CRASH (see client.py / server.py).
"""

from __future__ import annotations

import warnings

warnings.filterwarnings("ignore")


def run_pte(pte_bytes: bytes, inputs: list) -> list:
    """Load `pte_bytes` into the ExecuTorch runtime and run forward(*inputs).
    Returns a list of output tensors. May raise (Python) or hard-abort (native)."""
    from executorch.runtime import Runtime

    rt = Runtime.get()
    program = rt.load_program(pte_bytes)
    method = program.load_method("forward")
    out = method.execute([t.clone() for t in inputs])
    return list(out)

File: android_client/app/src/main/java/com/fuzzer/etclient/EtRunner.kt
package com.fuzzer.etclient

import org.pytorch.executorch.EValue
import org.pytorch.executorch.Module
import java.io.File

/**
 * Loads a .pte into the ExecuTorch runtime and runs forward(inputs). Mirrors
 * mobile/executor/runtime_run.py — the work a phone client does natively.
 *
 * A hard kernel failure is a NATIVE abort that takes the whole process down; the crash
 * breadcrumb in FuzzClient then reports the killed job as CRASH — the fuzzing signal we want.
 * scratchDir is PER-WORKER (see WorkerService) so job.pte never collides across processes.
 */
class EtRunner(private val scratchDir: File) {

    /** Write the .pte bytes to a file (Module.load needs a path), run, serialize raw outputs.
     *  The broker diffs these vs eager — the phone does no comparison. */
    fun run(pte: ByteArray, inputs: List<EValue>): List<Protocol.RawOut> {
        val f = File(scratchDir, "job.pte")
        f.writeBytes(pte)
        val module = Module.load(f.absolutePath)
        try {
            val outs: Array<EValue> = module.forward(*inputs.toTypedArray())
            return outs.filter { it.isTensor }.map { Tensors.etToRaw(it.toTensor()) }
        } finally {
            module.destroy()
        }
    }
}
